File: src/services/userActivitiesService.ts
import {
  collection,
  addDoc,
  doc,
  updateDoc,
  query,
  where,
  orderBy,
  limit,
  getDocs,
  Timestamp,
} from "firebase/firestore";
import { firestore } from "@/lib/firebase";

const USER_ACTIVITIES_COLLECTION = "userActivities";

export type ActivityType = "offer" | "reward" | "stamp" | "purchase" | "points";

export interface UserActivity {
  id: string;
  userId: string;
  type: ActivityType;
  title: string;
  description: string;
  storeName: string;
  storeId?: string;
  offerId?: string;
  redemptionId?: string;
  merchantId?: string;
  points?: number;
  status?: "pending" | "confirmed";
  createdAt: Date;
}

function toDate(value: unknown): Date {
  if (!value) return new Date();
  const v = value as { toDate?: () => Date; seconds?: number };
  if (typeof v?.toDate === "function") return v.toDate();
  if (typeof v?.seconds === "number") return new Date(v.seconds * 1000);
  return new Date();
}

/**
 * Registra atividade quando o usuário solicita uso de oferta
 */
export async function recordOfferRequested(params: {
  userId: string;
  redemptionId: string;
  offerId: string;
  offerTitle: string;
  storeId: string;
  storeName: string;
  merchantId: string;
}): Promise<void> {
  if (!firestore) return;
  try {
    await addDoc(collection(firestore, USER_ACTIVITIES_COLLECTION), {
      userId: params.userId,
      type: "offer",
      title: "Oferta solicitada",
      description: params.offerTitle,
      storeName: params.storeName,
      storeId: params.storeId,
      offerId: params.offerId,
      redemptionId: params.redemptionId,
      merchantId: params.merchantId,
      status: "pending",
      createdAt: Timestamp.now(),
    });
  } catch (e) {
    console.warn("[userActivitiesService] Erro ao registrar atividade:", e);
  }
}

/**
 * Atualiza atividade quando o lojista confirma o resgate
 */
export async function recordOfferConfirmed(redemptionId: string, merchantId: string): Promise<void> {
  if (!firestore) return;
  try {
    const q = query(
      collection(firestore, USER_ACTIVITIES_COLLECTION),
      where("redemptionId", "==", redemptionId)
    );
    const snap = await getDocs(q);
    for (const d of snap.docs) {
      await updateDoc(doc(firestore, USER_ACTIVITIES_COLLECTION, d.id), {
        status: "confirmed",
        title: "Oferta utilizada",
        updatedAt: Timestamp.now(),
      });
    }
  } catch (e) {
    console.warn("[userActivitiesService] Erro ao atualizar atividade:", e);
  }
}

/**
 * Busca atividades do usuário para a tela de histórico
 */
export async function getUserActivities(userId: string, limitCount = 50): Promise<UserActivity[]> {
  if (!firestore) return [];
  try {
    const q = query(
      collection(firestore, USER_ACTIVITIES_COLLECTION),
      where("userId", "==", userId),
      orderBy("createdAt", "desc"),
      limit(limitCount)
    );
    const snapshot = await getDocs(q);
    return snapshot.docs.map((d) => {
      const data = d.data();
      return {
        id: d.id,
        userId: String(data?.userId ?? ""),
        type: (data?.type as ActivityType) ?? "offer",
        title: String(data?.title ?? ""),
        description: String(data?.description ?? ""),
        storeName: String(data?.storeName ?? ""),
        storeId: data?.storeId ? String(data.storeId) : undefined,
        offerId: data?.offerId ? String(data.offerId) : undefined,
        redemptionId: data?.redemptionId ? String(data.redemptionId) : undefined,
        merchantId: data?.merchantId ? String(data.merchantId) : undefined,
        points: data?.points != null ? Number(data.points) : undefined,
        status: data?.status as "pending" | "confirmed" | undefined,
        createdAt: toDate(data?.createdAt),
      };
    });
  } catch (e) {
    console.error("[userActivitiesService] Erro ao buscar atividades:", e);
    return [];
  }
}

/**
 * Mescla atividades com resgates (para incluir resgates antigos sem atividade)
 */
export function mergeActivitiesWithRedemptions(
  activities: UserActivity[],
  redemptions: Array<{
    id: string;
    offerTitle: string;
    storeName: string;
    status: string;
    createdAt: Date;
  }>
): UserActivity[] {
  const redemptionIds = new Set(activities.map((a) => a.redemptionId).filter(Boolean));
  const fromRedemptions: UserActivity[] = redemptions
    .filter((r) => !redemptionIds.has(r.id))
    .map((r) => ({
      id: r.id,
      userId: "",
      type: "offer" as ActivityType,
      title: r.status === "confirmed" ? "Oferta utilizada" : "Oferta solicitada",
      description: r.offerTitle,
      storeName: r.storeName,
      redemptionId: r.id,
      status: (r.status === "confirmed" ? "confirmed" : "pending") as "pending" | "confirmed",
      createdAt: r.createdAt,
    }));
  return [...activities, ...fromRedemptions].sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
}

File: CartaoFidelidade/Services/UserActivitiesService.swift
//
//  UserActivitiesService.swift
//  CartaoFidelidade
//
//  Serviço para registrar e listar atividades do usuário.
//

import Foundation
import FirebaseFirestore
import FirebaseAuth

struct UserActivity: Identifiable {
    let id: String
    let userId: String
    let type: String // offer, reward, stamp, purchase, points
    let title: String
    let description: String
    let storeName: String
    let storeId: String?
    let offerId: String?
    let redemptionId: String?
    let merchantId: String?
    let points: Int?
    let status: String? // pending, confirmed
    let createdAt: Date
}

final class UserActivitiesService {
    static let shared = UserActivitiesService()
    private let db = Firestore.firestore()
    private let collectionName = "userActivities"

    private init() {}

    /// Registra atividade quando o usuário solicita uso de oferta
    func recordOfferRequested(
        userId: String,
        redemptionId: String,
        offerId: String,
        offerTitle: String,
        storeId: String,
        storeName: String,
        merchantId: String
    ) {
        let data: [String: Any] = [
            "userId": userId,
            "type": "offer",
            "title": "Oferta solicitada",
            "description": offerTitle,
            "storeName": storeName,
            "storeId": storeId,
            "offerId": offerId,
            "redemptionId": redemptionId,
            "merchantId": merchantId,
            "status": "pending",
            "createdAt": Timestamp(date: Date()),
        ]
        db.collection(collectionName).addDocument(data: data) { error in
            if let error = error {
                print("❌ [UserActivitiesService] Erro ao registrar atividade: \(error.localizedDescription)")
            }
        }
    }

    /// Atualiza atividade quando o lojista confirma o resgate
    func recordOfferConfirmed(redemptionId: String) {
        db.collection(collectionName)
            .whereField("redemptionId", isEqualTo: redemptionId)
            .getDocuments { [weak self] snapshot, error in
                guard let self = self, error == nil, let docs = snapshot?.documents else { return }
                for doc in docs {
                    doc.reference.updateData([
                        "status": "confirmed",
                        "title": "Oferta utilizada",
                        "updatedAt": Timestamp(date: Date()),
                    ]) { err in
                        if let err = err {
                            print("❌ [UserActivitiesService] Erro ao atualizar: \(err.localizedDescription)")
                        }
                    }
                }
            }
    }

    /// Busca atividades do usuário
    func getUserActivities(userId: String, limitCount: Int = 50) async throws -> [UserActivity] {
        let snapshot = try await db.collection(collectionName)
            .whereField("userId", isEqualTo: userId)
            .order(by: "createdAt", descending: true)
            .limit(to: limitCount)
            .getDocuments()

        return snapshot.documents.compactMap { doc -> UserActivity? in
            let data = doc.data()
            guard let createdAt = (data["createdAt"] as? Timestamp)?.dateValue() else { return nil }
            return UserActivity(
                id: doc.documentID,
                userId: data["userId"] as? String ?? "",
                type: data["type"] as? String ?? "offer",
                title: data["title"] as? String ?? "",
                description: data["description"] as? String ?? "",
                storeName: data["storeName"] as? String ?? "",
                storeId: data["storeId"] as? String,
                offerId: data["offerId"] as? String,
                redemptionId: data["redemptionId"] as? String,
                merchantId: data["merchantId"] as? String,
                points: data["points"] as? Int,
                status: data["status"] as? String,
                createdAt: createdAt
            )
        }
    }
}
